src/common/task.cc

```diff
@@ -348,7 +359,80 @@ const std::unordered_map<std::string, double> TaskSpec_get_required_resources(
   return map_from_flatbuf(*message->required_resources());
 }
 
-bool TaskSpec_is_dependent_on(TaskSpec *spec, ObjectID object_id) {
+TaskSpec *TaskSpec_copy(TaskSpec *spec, int64_t task_spec_size) {
+  TaskSpec *copy = (TaskSpec *) malloc(task_spec_size);
+  memcpy(copy, spec, task_spec_size);
+  return copy;
+}
+
+void TaskSpec_free(TaskSpec *spec) {
+  free(spec);
+}
+
+TaskExecutionSpec::TaskExecutionSpec(
+    const std::vector<ObjectID> execution_dependencies,
+    TaskSpec *spec,
+    int64_t task_spec_size) {
+  execution_dependencies_ = execution_dependencies;
+  task_spec_size_ = task_spec_size;
+  TaskSpec *spec_copy = new TaskSpec[task_spec_size_];
+  memcpy(spec_copy, spec, task_spec_size);
+  spec_ = std::unique_ptr<TaskSpec[]>(spec_copy);
+}
+
+TaskExecutionSpec::TaskExecutionSpec(TaskExecutionSpec *other) {
+  execution_dependencies_ = other->execution_dependencies_;
+  task_spec_size_ = other->task_spec_size_;
+  TaskSpec *spec_copy = new TaskSpec[task_spec_size_];
+  memcpy(spec_copy, other->spec_.get(), task_spec_size_);
+  spec_ = std::unique_ptr<TaskSpec[]>(spec_copy);
+}
+
+std::vector<ObjectID> TaskExecutionSpec::ExecutionDependencies() {
+  return execution_dependencies_;
+}
+
+int64_t TaskExecutionSpec::SpecSize() {
+  return task_spec_size_;
+}
+
+TaskSpec *TaskExecutionSpec::Spec() {
+  return spec_.get();
+}
+
+int64_t TaskExecutionSpec::NumDependencies() {
+  TaskSpec *spec = Spec();
+  int64_t num_dependencies = TaskSpec_num_args(spec);
```

Comment:
doesn't this count non ObjectID arguments?

Reply:
Yes, all "dependencies" are included right now. Alternatively, we could change `TaskExecutionSpec` to only include `execution_dependencies`. I think it's a bit cleaner, though, to have all dependencies together, to avoid cases where we accidentally only consider one dependency type or the other.